File: CSharp/LeetCode/Solutions/1137-nth-tribonacci-number.cs
using LeetCode.Solutions.Shared;

namespace LeetCode.Solutions;

internal class Solution1137 : IRunProgram
{
    public void Run()
    {
        Tribonacci(2).Print();
        Tribonacci(22).Print();
    }
    public int Tribonacci(int n)
    {
        int[] mem = new int[n];
        return Recursive(n, mem);
    }

    private int Recursive(int n, int[] mem)
    {
        if (n <= 0)
        {
            return 0;
        }
        else if (n == 1)
        {
            return 1;
        }
        else if (mem[n - 1] != 0)
        {
            return mem[n - 1];
        }
        else
        {
            mem[n - 1] = Recursive(n - 1, mem) + Recursive(n - 2, mem) + Recursive(n - 3, mem);
            return mem[n - 1];
        }
    }
}

File: Python/1137-n-th-tribonacci-number.py
class Solution:
    def tribonacci(self, n: int) -> int:
        mem = [0] * (n + 1)
        return self.rec(n, mem)
    def rec(self, n, mem) :
        if n <= 0 :
            return 0
        elif n <= 2 :
            return 1
        elif mem[n - 1] == 0 :
            mem[n - 1] = self.rec(n - 3, mem) +  self.rec(n - 2, mem) +  self.rec(n - 1, mem)
        return mem[n - 1]
        